nemo_curator/classifiers/aegis.py

```diff
@@ -0,0 +1,246 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+
+os.environ["RAPIDS_NO_INITIALIZE"] = "1"
+os.environ["DASK_DATAFRAME__QUERY_PLANNING"] = "False"
+from dataclasses import dataclass
+from functools import lru_cache
+from typing import Optional, Union
+
+import cudf
+import torch
+import torch.nn as nn
+from crossfit import op
+from crossfit.backend.torch.hf.model import HFModel
+from peft import PeftModel
+from transformers import AutoConfig, AutoModelForCausalLM, AutoTokenizer
+
+from nemo_curator.classifiers.base import DistributedDataClassifier
+from nemo_curator.datasets import DocumentDataset
+from nemo_curator.utils.aegis_utils import format_aegis
+
+
+@dataclass
+class AegisConfig:
+    peft_model_name_or_path: str
+    token: Optional[Union[str, bool]] = None
+    pretrained_model_name_or_path: str = "meta-llama/LlamaGuard-7b"
+    dtype: torch.dtype = torch.bfloat16
+    max_length: int = 4096
+
+
+AEGIS_LABELS = [
+    "unknown",
+    "safe",
+    "O1",
+    "O2",
+    "O3",
+    "O4",
+    "O5",
+    "O6",
+    "O7",
+    "O8",
+    "O9",
+    "O10",
+    "O11",
+    "O12",
+    "O13",
+]
+
+
+class AegisModel(nn.Module):
+    def __init__(
+        self,
+        pretrained_model_name_or_path: str,
+        peft_model_name_or_path: str,
+        dtype: torch.dtype,
+        token: str,
+    ):
+        super().__init__()
+        base_model = AutoModelForCausalLM.from_pretrained(
+            pretrained_model_name_or_path, torch_dtype=dtype, token=token
+        )
+        self.model = PeftModel.from_pretrained(base_model, peft_model_name_or_path)
+
+    @torch.no_grad()
+    def forward(self, batch):
+        response = self.model.generate(
+            **batch,
+            max_new_tokens=100,
+            pad_token_id=0,
+        )
```

Comment:
Question: Is there  a reason we don't use `autocast` here ?

Reply:
LlamaGuard 7B is already in [bfloat16](https://huggingface.co/meta-llama/LlamaGuard-7b/blob/main/config.json#L22). I'm not too familiar with autocast, but I figured it was not needed since the model is already in a lower precision format. I totally could be wrong though.

Reply:
Fair enough, I think in that case its not really required.